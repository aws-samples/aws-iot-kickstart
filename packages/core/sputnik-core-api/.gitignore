lib/
generated/
schema.graphql
!bin/**
