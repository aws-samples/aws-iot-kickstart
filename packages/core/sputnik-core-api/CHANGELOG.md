# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2020-08-09)


### Features

* implement batch device deploy api/infra ([4a2ae89](https://git-codecommit.us-west-2.amazonaws.com/v1/repos/Deathstar/commits/4a2ae89f7f3f734521cf738683fcaddd2454dd78))
* **sputnik-core-api:** copy graphql operation files from ui in core api ([2ac85c0](https://git-codecommit.us-west-2.amazonaws.com/v1/repos/Deathstar/commits/2ac85c0a1e045bf416a0c90d22afddc1f0778745))
* **sputnik-core-api:** initial core api implementation ([38132c7](https://git-codecommit.us-west-2.amazonaws.com/v1/repos/Deathstar/commits/38132c72e9d7d2e5ff7309f004580ec8e78c930d))
