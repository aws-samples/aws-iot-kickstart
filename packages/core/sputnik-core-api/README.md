# `@deathstar/sputnik-core-api`

> TODO: description

Sputnik core api schema, types, operation, etc

## Usage

```
import api from '@deathstar/sputnik-core-api';

// TODO: Add usage details
```
