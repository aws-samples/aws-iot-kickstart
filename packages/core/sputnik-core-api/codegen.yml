overwrite: true
schema:
  [
    './src/schema/appsync.graphql',
    './lib/schema.graphql',
  ]
documents: 'src/operation/**/*.graphql'
generates:
  generated/types.ts:
    plugins:
      - 'typescript'
  generated/apollo-angular.ts:
    preset: import-types
    presetConfig:
      typesPath: ./types
    plugins:
      - 'typescript-operations'
      - 'typescript-apollo-angular'
    config:
      serviceName: SputnikApiService
      sdkClass: true
  generated/graphql-request.ts:
    preset: import-types
    presetConfig:
      typesPath: ./types
    plugins:
      - 'typescript-operations'
      - 'typescript-graphql-request'
