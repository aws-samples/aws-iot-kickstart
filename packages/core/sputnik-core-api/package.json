{
  "name": "@deathstar/sputnik-core-api",
  "description": "Sputnik core api schema, types, operation, etc",
  "version": "0.0.0-beta.0",
  "license": "Apache-2.0",
  "publishConfig": {
    "access": "public",
    "tag": "beta"
  },
  "author": "Jeremy Jonas <jay.jonas@hotmail.com>",
  "contributors": [
    "Timothee Cruse <timothee.cruse@gmail.com>",
    "Jeremy Jonas <jay.jonas@hotmail.com>",
    "Greg Varga <contact@gergelyvarga.com>",
    "Mirash Gjolaj <mirash.gjolaj@gmail.com>"
  ],
  "directories": {
    "lib": "lib",
    "test": "test"
  },
  "files": [
    "lib",
    "schema.graphql"
  ],
  "main": "lib/index.js",
  "typings": "lib/types.d.ts",
  "scripts": {
    "clean": "rm -rf lib generated",
    "prebuild": "yarn clean && yarn generate",
    "build": "tsc --project tsconfig.json",
    "postbuild": "cat src/schema/appsync.graphql > lib/schema.dev.graphql && cat lib/schema.graphql >> lib/schema.dev.graphql",
    "test": "jest",
    "generate": "yarn generate:schema && yarn generate:code && yarn generate:src",
    "generate:src": "cp src/schema-file.ts generated/schema-file.ts",
    "generate:schema": "ts-node ./bin/generate-schema.js",
    "generate:code": "graphql-codegen --config codegen.yml",
    "prewatch": "yarn clean && yarn generate:schema",
    "watch:schema": "nodemon -e ts,graphql -w src/schema -x yarn generate:schema",
    "watch:code": "yarn generate:code --watch",
    "watch": "run-p -c watch:*"
  },
  "dependencies": {
    "apollo-angular": "^1.10.0",
    "apollo-cache": "^1.3.5",
    "apollo-client": "^2.6.10",
    "graphql": "^15.3.0",
    "graphql-request": "^2.0.0"
  },
  "devDependencies": {
    "@graphql-codegen/cli": "^1.17.6",
    "@graphql-codegen/import-types-preset": "^1.17.6",
    "@graphql-codegen/typescript": "^1.17.6",
    "@graphql-codegen/typescript-apollo-angular": "^1.17.0",
    "@graphql-codegen/typescript-graphql-request": "^1.17.6",
    "@graphql-codegen/typescript-operations": "^1.17.6",
    "@graphql-codegen/typescript-react-apollo": "^1.17.6",
    "appsync-schema-converter": "^0.1.3",
    "graphql-tools": "^6.0.12",
    "graphql.macro": "^1.4.2",
    "nodemon": "^2.0.4",
    "ts-node": "^8.10.2"
  }
}
