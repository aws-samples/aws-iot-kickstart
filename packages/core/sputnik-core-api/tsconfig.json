{
  "extends": "../../../configs/workspace/tsconfig.json",
  "compilerOptions": {
    "noEmit": false,
    "emitDeclarationOnly": false,
    "skipLibCheck": true,
    "experimentalDecorators": true,
    "declaration": true,
    "declarationDir": "./lib",
    "rootDir": "./generated",
    "outDir": "./lib"
  },
  "compileOnSave": true,
  "exclude": ["node_modules", "dist"],
  "include": ["generated"],
}
