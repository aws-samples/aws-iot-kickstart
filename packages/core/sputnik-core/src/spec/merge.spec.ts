import { SpecDefinition } from '@deathstar/sputnik-core-api'
import { mergeSpecs } from './merge'

// TODO: move this to global
type JestExpect = <R>(actual: R) => jest.Matchers<R> & jasmine.Matchers<R>;
declare const expect: JestExpect;

describe('merge', () => {
	test('should merge all specs', () => {
		const deviceSpec = createSpec('device')
		const deviceTypeSpec = createSpec('device-type')
		const deviceBlueprintSpec = createSpec('device-blueprint')

		const spec = mergeSpecs(deviceBlueprintSpec, deviceTypeSpec, deviceSpec)

		expect(spec).toMatchSnapshot('merge-all')
	})

	test('should override', () => {
		const deviceSpec = createSpec('device')
		const deviceTypeSpec = createSpec('device-type')
		const deviceBlueprintSpec = createSpec('device-blueprint')

		if (deviceBlueprintSpec.FunctionDefinitionVersion && deviceSpec.FunctionDefinitionVersion) {
			deviceSpec.FunctionDefinitionVersion.Functions[0].FunctionArn = 'override-device'
			deviceBlueprintSpec.FunctionDefinitionVersion.Functions[0].FunctionArn = 'override-device'
		}

		if (deviceTypeSpec.ConnectorDefinitionVersion && deviceSpec.ConnectorDefinitionVersion) {
			deviceSpec.ConnectorDefinitionVersion.Connectors[0].ConnectorArn = 'override-device'
			deviceTypeSpec.ConnectorDefinitionVersion.Connectors[0].ConnectorArn = 'override-device'
		}

		const spec = mergeSpecs(deviceBlueprintSpec, deviceTypeSpec, deviceSpec)

		expect(spec.FunctionDefinitionVersion?.Functions.length).toBe(2)
		expect(spec.FunctionDefinitionVersion?.Functions[1]).toEqual(deviceSpec.FunctionDefinitionVersion?.Functions[0])

		expect(spec.ConnectorDefinitionVersion?.Connectors.length).toBe(2)
		expect(spec.ConnectorDefinitionVersion?.Connectors[1]).toEqual(deviceSpec.ConnectorDefinitionVersion?.Connectors[0])
	})
})

function createSpec (prefix: string): SpecDefinition {
	return {
		CoreDefinitionVersion: {
			Cores: [
				{
					CertificateArn: `${prefix}-CertificateArn`,
					Id: `${prefix}-Id`,
					ThingArn: `${prefix}-ThingArn`,
				}
			]
		},
		ConnectorDefinitionVersion: {
			Connectors: [
				{
					ConnectorArn: `${prefix}-ConnectorArn`,
					Id: `${prefix}-Id`,
					Parameters: `${prefix}-Parameters`,
				}
			]
		},
		DeviceDefinitionVersion: {
			Devices: [
				{
					CertificateArn: `${prefix}-CertificateArn`,
					Id: `${prefix}-Id`,
					ThingArn: `${prefix}-ThingArn`,
				}
			]
		},
		FunctionDefinitionVersion: {
			DefaultConfig: {
				Execution: {
					RunAs: {
						Gid: 111,
						Uid: 111,
					},
				}
			},
			Functions: [
				{
					FunctionArn: `${prefix}-FunctionArn`,
					FunctionConfiguration: {
						EncodingType: `${prefix}-EncodingType`,
					},
					Id: `${prefix}-Id`,
				}
			]
		},
		LoggerDefinitionVersion: {
			Loggers: [
				{
					Id: `${prefix}-Id`,
					Component: `${prefix}-Component`,
					Level: `${prefix}-Level`,
					Type: `${prefix}-Type`,
				}
			]
		},
		ResourceDefinitionVersion: {
			Resources: [
				{
					Id: `${prefix}-Id`,
					Name: `${prefix}-Name`,
					ResourceDataContainer: {
						LocalDeviceResourceData: {
							SourcePath: `${prefix}-SourcePath`,
						},
						LocalVolumeResourceData: {
							SourcePath: `${prefix}-SourcePath`,
							DestinationPath: `${prefix}-DestinationPath`,
						}
					}
				}
			]
		},
		SubscriptionDefinitionVersion: {
			Subscriptions: [
				{
					Id: `${prefix}-Id`,
					Source: `${prefix}-Source`,
					Subject: `${prefix}-Subject`,
					Target: `${prefix}-Target`,
				}
			]
		}
	}
}
