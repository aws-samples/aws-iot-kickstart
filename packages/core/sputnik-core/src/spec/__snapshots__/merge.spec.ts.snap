// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge should merge all specs: merge-all 1`] = `
Object {
  "ConnectorDefinitionVersion": Object {
    "Connectors": Array [
      Object {
        "ConnectorArn": "device-blueprint-ConnectorArn",
        "Id": "device-blueprint-Id",
        "Parameters": "device-blueprint-Parameters",
      },
      Object {
        "ConnectorArn": "device-type-ConnectorArn",
        "Id": "device-type-Id",
        "Parameters": "device-type-Parameters",
      },
      Object {
        "ConnectorArn": "device-ConnectorArn",
        "Id": "device-Id",
        "Parameters": "device-Parameters",
      },
    ],
  },
  "CoreDefinitionVersion": Object {
    "Cores": Array [
      Object {
        "CertificateArn": "device-CertificateArn",
        "Id": "device-Id",
        "ThingArn": "device-ThingArn",
      },
    ],
  },
  "DeviceDefinitionVersion": Object {
    "Devices": Array [
      Object {
        "CertificateArn": "device-blueprint-CertificateArn",
        "Id": "device-blueprint-Id",
        "ThingArn": "device-blueprint-ThingArn",
      },
      Object {
        "CertificateArn": "device-type-CertificateArn",
        "Id": "device-type-Id",
        "ThingArn": "device-type-ThingArn",
      },
      Object {
        "CertificateArn": "device-CertificateArn",
        "Id": "device-Id",
        "ThingArn": "device-ThingArn",
      },
    ],
  },
  "FunctionDefinitionVersion": Object {
    "DefaultConfig": Object {
      "Execution": Object {
        "RunAs": Object {
          "Gid": 111,
          "Uid": 111,
        },
      },
    },
    "Functions": Array [
      Object {
        "FunctionArn": "device-blueprint-FunctionArn",
        "FunctionConfiguration": Object {
          "EncodingType": "device-blueprint-EncodingType",
        },
        "Id": "device-blueprint-Id",
      },
      Object {
        "FunctionArn": "device-type-FunctionArn",
        "FunctionConfiguration": Object {
          "EncodingType": "device-type-EncodingType",
        },
        "Id": "device-type-Id",
      },
      Object {
        "FunctionArn": "device-FunctionArn",
        "FunctionConfiguration": Object {
          "EncodingType": "device-EncodingType",
        },
        "Id": "device-Id",
      },
    ],
  },
  "LoggerDefinitionVersion": Object {
    "Loggers": Array [
      Object {
        "Component": "device-blueprint-Component",
        "Id": "device-blueprint-Id",
        "Level": "device-blueprint-Level",
        "Type": "device-blueprint-Type",
      },
      Object {
        "Component": "device-type-Component",
        "Id": "device-type-Id",
        "Level": "device-type-Level",
        "Type": "device-type-Type",
      },
      Object {
        "Component": "device-Component",
        "Id": "device-Id",
        "Level": "device-Level",
        "Type": "device-Type",
      },
    ],
  },
  "ResourceDefinitionVersion": Object {
    "Resources": Array [
      Object {
        "Id": "device-blueprint-Id",
        "Name": "device-blueprint-Name",
        "ResourceDataContainer": Object {
          "LocalDeviceResourceData": Object {
            "SourcePath": "device-blueprint-SourcePath",
          },
          "LocalVolumeResourceData": Object {
            "DestinationPath": "device-blueprint-DestinationPath",
            "SourcePath": "device-blueprint-SourcePath",
          },
        },
      },
      Object {
        "Id": "device-type-Id",
        "Name": "device-type-Name",
        "ResourceDataContainer": Object {
          "LocalDeviceResourceData": Object {
            "SourcePath": "device-type-SourcePath",
          },
          "LocalVolumeResourceData": Object {
            "DestinationPath": "device-type-DestinationPath",
            "SourcePath": "device-type-SourcePath",
          },
        },
      },
      Object {
        "Id": "device-Id",
        "Name": "device-Name",
        "ResourceDataContainer": Object {
          "LocalDeviceResourceData": Object {
            "SourcePath": "device-SourcePath",
          },
          "LocalVolumeResourceData": Object {
            "DestinationPath": "device-DestinationPath",
            "SourcePath": "device-SourcePath",
          },
        },
      },
    ],
  },
  "SubscriptionDefinitionVersion": Object {
    "Subscriptions": Array [
      Object {
        "Id": "device-blueprint-Id",
        "Source": "device-blueprint-Source",
        "Subject": "device-blueprint-Subject",
        "Target": "device-blueprint-Target",
      },
      Object {
        "Id": "device-type-Id",
        "Source": "device-type-Source",
        "Subject": "device-type-Subject",
        "Target": "device-type-Target",
      },
      Object {
        "Id": "device-Id",
        "Source": "device-Source",
        "Subject": "device-Subject",
        "Target": "device-Target",
      },
    ],
  },
}
`;
